File: src/21_ReorderArray/21_ReorderArray.py
class Solution:
    def exchange(self, nums: list[int]) -> list[int]:
        i, j = 0, len(nums) - 1
        while i < j:
            while i < j and (nums[i] & 1) == 1:
                i += 1
            while i < j and (nums[j] & 1) == 0:
                j -= 1
            nums[i], nums[j] = nums[j], nums[i]
        return nums


class Test:
    def __init__(self) -> None:
        self.solution = Solution()

    def checkResult(self, testName: str, nums: list[int]) -> None:
        print(f"==== {testName} ====")
        print("original:")
        print(nums)

        result = self.solution.exchange(nums)
        print("exchanged:")
        print(result)
        print("\n")

    def Test1(self):
        nums = [1, 2, 3, 4, 5, 6, 7]
        self.checkResult("Test1", nums)

    def Test2(self):
        nums = [2, 4, 6, 1, 3, 5, 7]
        self.checkResult("Test2", nums)

    def Test3(self):
        nums = [1, 3, 5, 7, 2, 4, 6]
        self.checkResult("Test3", nums)

    def Test4(self):
        nums = [1]
        self.checkResult("Test4", nums)

    def Test5(self):
        nums = [2]
        self.checkResult("Test5", nums)

    def Test6(self):
        nums = []
        self.checkResult("Test6", nums)

    def runTest(self):
        self.Test1()
        self.Test2()
        self.Test3()
        self.Test4()
        self.Test5()
        self.Test6()


if __name__ == "__main__":
    test = Test()
    test.runTest()

File: src/21_ReorderArray/21_ReorderArray.cpp

// 面试题21：调整数组顺序使奇数位于偶数前面
// 题目：输入一个整数数组，实现一个函数来调整该数组中数字的顺序，使得所有
// 奇数位于数组的前半部分，所有偶数位于数组的后半部分。

#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

class Solution {
   public:
    // 双指针：i从左向右找偶数，j从右向左找奇数，
    // 然后交换两指针
    vector<int> exchange( vector<int>& nums ) {
        int i = 0, j = nums.size() - 1;
        while ( i < j ) {
            while ( i < j && ( nums[ i ] & 1 ) == 1 ) i++;
            while ( i < j && ( nums[ j ] & 1 ) == 0 ) j--;
            std::swap( nums[ i ], nums[ j ] );
        }
        return nums;
    }
};

class Test {
   public:
    void checkResult( const string testName, vector<int>& nums ) {
        cout << "==== " << testName << " ====\n";
        cout << "original:\n";
        printArray( nums );

        vector<int> result = solution.exchange( nums );
        cout << "exchanged:\n";
        printArray( result );
        cout << endl;
    }

    void Test1() {
        vector<int> nums{ 1, 2, 3, 4, 5, 6, 7 };
        checkResult( "Test1", nums );
    }

    void Test2() {
        vector<int> nums{ 2, 4, 6, 1, 3, 5, 7 };
        checkResult( "Test2", nums );
    }

    void Test3() {
        vector<int> nums{ 1, 3, 5, 7, 2, 4, 6 };
        checkResult( "Test3", nums );
    }

    void Test4() {
        vector<int> nums{ 1 };
        checkResult( "Test4", nums );
    }

    void Test5() {
        vector<int> nums{ 2 };
        checkResult( "Test5", nums );
    }

    void Test6() {
        vector<int> nums;
        checkResult( "Test6", nums );
    }

    void runTest() {
        Test1();
        Test2();
        Test3();
        Test4();
        Test5();
        Test6();
    }

   private:
    Solution solution;

    void printArray( const vector<int>& nums ) {
        cout << "[ ";
        for ( int num : nums ) cout << num << " ";
        cout << "]\n";
    }
};

int main() {
    Test test;
    test.runTest();
    return 0;
}